### ISSUE TITLE
bug: wrong get_caller_address() in constructor when using cairo-test

### ISSUE THREAD

**Issue by credence0x** [starkware-libs/cairo#3844]
# Bug Report

When using `cairo-test`, `starknet::get_caller_address()`  might not return the right address when called in a constructor.

**Cairo version:**
v2.0.0

 https://github.com/starkware-libs/cairo/tree/v2.0.0

**Current behavior:**

When a you use `deploy_syscall` from a contract, the caller address in the constructor of the CALLED/ DEPLOYED contract using the syscall will be the caller address of the CALLING contract

**Expected behavior:**

The caller address in the constructor of the CALLED/ DEPLOYED contract should be the CALLING contract's address

**Steps to reproduce:**

Check out snippet below

**Related code:**

```
use starknet::ClassHash;

#[starknet::contract]
mod ContractB {
    use debug::PrintTrait;

    #[storage]
    struct Storage {}

    #[constructor]
    fn constructor(ref self: ContractState) {
        // print deployer address
        starknet::get_caller_address().print();
    }
}


#[starknet::interface]
trait IContractA<TContractState> {
    fn start(self: @TContractState, class_hash_b: ClassHash);
}

#[starknet::contract]
mod ContractA {
    use starknet::ClassHash;
    use starknet::syscalls::deploy_syscall;
    use array::{ArrayTrait, SpanTrait};
    use debug::PrintTrait;


    #[storage]
    struct Storage {}

    #[constructor]
    fn constructor(ref self: ContractState) {}

    #[external(v0)]
    impl ContratAImpl of super::IContractA<ContractState> {
        fn start(self: @ContractState, class_hash_b: ClassHash) {
            let mut calldata = Default::default();
            // when deploying ContractB through this contract, 
            // the constructor of ContractB will be called and it is 
            // expected to print ContractB's caller address, which is 
            // the address of this contract. It will however print the
            // address of the caller of this function

            let syscall_result = deploy_syscall(
                class_hash_b, 0, calldata.span(), false
            );

            // print this contract's address
            starknet::get_contract_address().print();
        }
    }
}

#[cfg(test)]
mod test_caller_addresses {
    use starknet::syscalls::deploy_syscall;
    use array::{ArrayTrait, SpanTrait};
    use super::{ContractA, ContractB};
    use super::{IContractA, IContractADispatcher, IContractADispatcherTrait};
    use traits::{Into, TryInto};
    use option::OptionTrait;
    use result::ResultTrait;

    #[test]
    #[available_gas(300000)]
    fn do_test(){
        let mut calldata = Default::default();
        let (contract_a_address, _) = deploy_syscall(
            ContractA::TEST_CLASS_HASH.try_into().unwrap(),
            0,
            calldata.span(),
            false
        ).unwrap();
        let contract_a = IContractADispatcher { 
            contract_address: contract_a_address 
        };
        contract_a.start(ContractB::TEST_CLASS_HASH.try_into().unwrap());
        // check the logs for the output. It is expected that the same address
        // will be printed twice - once by ContractA and once by ContractB 
        // but it is not the case.
    }
}
```

**Other information:**

**Comment by orizi** [comment]
As far as i know this is fixed in 2.1.0.

**Comment by credence0x** [comment]
> As far as i know this is fixed in 2.1.0.

I just tried running the snippet in the description with 
`$scarb --version`
```
scarb 0.6.0 (532d8e349 2023-08-07)
cairo: 2.1.0 (https://crates.io/crates/cairo-lang-compiler/2.1.0)
```
It produced the same result

**Comment by credence0x** [comment]
> As far as i know this is fixed in 2.1.0.

I just tried running the snippet in the description with 
`$scarb --version`
```
scarb 0.6.0 (532d8e349 2023-08-07)
cairo: 2.1.0 (https://crates.io/crates/cairo-lang-compiler/2.1.0)
```
It produced the same result

**Comment by orizi** [comment]
you are correct - got confused with a older bug - fixing now.
